flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/BucketPartitioner.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink;
+
+import java.util.Optional;
+import org.apache.flink.api.common.functions.Partitioner;
+import org.apache.iceberg.PartitionField;
+import org.apache.iceberg.PartitionSpec;
+
+/**
+ * This partitioner will redirect elements to writers deterministically so that each writer only
+ * targets 1 bucket. If the number of writers > number of buckets each partitioner will keep a state
+ * of multiple writers per bucket as evenly as possible, and will round-robin the requests across
+ * them.
+ */
+public class BucketPartitioner implements Partitioner<Integer> {
+
+  private final Integer maxBuckets;
+
+  private final int[] currentWriterOffset;
+
+  BucketPartitioner(PartitionSpec partitionSpec) {
+    // The current implementation redirects to writers based on the _FIRST_ bucket found
+    Optional<PartitionField> bucket =
+        partitionSpec.fields().stream()
+            .filter(f -> f.transform().dedupName().contains("bucket"))
+            .findFirst();
+
+    if (!bucket.isPresent()) {
```

Comment:
OK, we can fail here in case this partitioner is accessed directly. And have the `FlinkSink.distributeDataStream` only use this strategy if only 1 bucket is defined in the partition spec.